### PR TITLE
Added LoginWithAmazon and Google Sign-in methods

### PR DESCRIPTION
**Pull request by omenocal**
Amazon docs here:
https://developer.amazon.com/docs/login-with-amazon/customer-profile.html

Google Assistant docs here:
https://developers.google.com/actions/identity/google-sign-in

### CODE REVIEW

**Comment by omenocal** [comment]
@armonge done. Also replaced some places where it said: `js::method` to `js::function` as this was having issues when being created with sphinx

File: src/VoxaEvent.ts [hunk 1]
```diff
@@ -87,6 +90,44 @@ export abstract class VoxaEvent implements IVoxaEvent {
     this.initLogger(logOptions);
   }
 
+  public async getUserInformation(): Promise<any> {
+    if (_.has(this, "google")) {
```
Comment (armonge) [inline src/VoxaEvent.ts]: why not `this.platform.name === 'dialogflow'` ?

File: src/VoxaEvent.ts [hunk 2]
```diff
@@ -87,6 +90,44 @@ export abstract class VoxaEvent implements IVoxaEvent {
...
+      return this.getUserInformationWithGoogle();
+    }
+
+    return await this.getUserInformationWithLWA();
+  }
+
+  public getUserInformationWithGoogle(): any {
```
Comment (armonge) [inline src/VoxaEvent.ts]: please specify the the return type, create an interface if none is available

File: src/VoxaEvent.ts [hunk 3]
```diff
@@ -87,6 +90,44 @@ export abstract class VoxaEvent implements IVoxaEvent {
...
+    const result: any = jwt.decode(this.user.idToken);
```
Comment (armonge) [inline src/VoxaEvent.ts]: `this.user.idToken` ??

File: src/VoxaEvent.ts [hunk 4]
```diff
@@ -87,6 +90,44 @@ export abstract class VoxaEvent implements IVoxaEvent {
...
+    result.givenName = result.given_name;
+
+    delete result.email_verified;
+    delete result.family_name;
+    delete result.given_name;
+
+    return result;
+  }
+
+  public async getUserInformationWithLWA(): Promise<any> {
```
Comment (armonge) [inline src/VoxaEvent.ts]: please create an interface for this return type, would be great if both methods share that BTW

File: src/VoxaEvent.ts [hunk 5]
```diff
@@ -87,6 +90,44 @@ export abstract class VoxaEvent implements IVoxaEvent {
...
+    const httpOptions: any = {
+      json: true,
+      method: "GET",
+      uri: `https://api.amazon.com/user/profile?access_token=${this.user.accessToken}`,
```
Comment (armonge) [inline src/VoxaEvent.ts]: what happens if the user has no accessToken and the developer calls this method? should it return a RuntimeException or something like that?

File: src/VoxaEvent.ts [hunk 6]
```diff
@@ -87,6 +90,44 @@ export abstract class VoxaEvent implements IVoxaEvent {
...
+    };
+
+    const result: any = await rp(httpOptions);
```
Comment (armonge) [inline src/VoxaEvent.ts]: create an interface for what this http call returns

File: src/VoxaEvent.ts (hunk 3 above)
Comment (omenocal) [reply src/VoxaEvent.ts]: It's the variable that has the JSON Web Token with the user's Google information. It's not an accessToken

Comment (armonge) [reply src/VoxaEvent.ts]: yes but, from the dialogflow implementation for the user object, seems like this is not correct https://github.com/VoxaAI/voxa/blob/v3/src/platforms/dialogflow/DialogFlowEvent.ts#L91

perhaps you're looking for `voxaEvent.dialogflow.conv.user.idToken` ?

**Comment by omenocal** [comment]
@armonge done

File: src/index.ts [hunk 7]
```diff
@@ -7,7 +7,7 @@
 
 export { Tell, Say, SayP, Ask, Reprompt } from "./directives";
 export { VoxaPlatform } from "./platforms";
-export { IVoxaEvent, IVoxaIntentEvent, IVoxaIntent } from "./VoxaEvent";
+export { IVoxaEvent, IVoxaIntentEvent, IVoxaIntent, VoxaEvent } from "./VoxaEvent";
```
[the remark this answers is not in the document]
Comment (omenocal) [reply src/index.ts]: @armonge done

File: test/alexa/AlexaEvent.spec.ts [hunk 8]
```diff
@@ -69,3 +73,57 @@ describe("AlexaEvent", () => {
...
+    const userDetails = await alexaEvent.getUserInformation();
+    expect(userDetails).to.deep.equal(lwaResult);
+  });
+
+  it("should throw an error when accessToken is empty", async () => {
+    const rawEvent = rb.getLaunchRequest();
+    const alexaEvent = new AlexaEvent(rawEvent) as VoxaEvent;
+    alexaEvent.platform = alexaSkill;
+
+    try {
```
Comment (armonge) [inline test/alexa/AlexaEvent.spec.ts]: maybe also add something to actually make sure the exception is thrown? right now we're only testing that if the exception is thrown the message is what we expect


```suggestion
    let exceptionWasThrown = false;
    try {
      await alexaEvent.getUserInformation();
    } catch (err) {
      exceptionWasThrown = true;
      expect(err.message).to.equal("this.user.accessToken is empty");
    }
    expect(exceptionWasThrown).to.be.true;
```

Comment (omenocal) [reply test/alexa/AlexaEvent.spec.ts]: @armonge done

File: docs/google-sign-in.rst [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+To perform account linking with Google Sign-In, you ask the user to give consent to access their Google profile. You then use the information in their profile, for example their email address, to identify the user in your system. Check out this `link <https://developers.google.com/actions/identity/google-sign-in>`_ for more information.
+
+With Voxa, you can ask for the user's full name like this:
+
+.. code-block:: javascript
+
+  app.onIntent('ProfileIntent', async (dialogFlowEvent: DialogFlowEvent) => {
+    const userInfo = await dialogFlowEvent.getUserInformation();
+
+    dialogFlowEvent.model.email = userInfo.email;
```
Comment (rmberrios) [inline docs/google-sign-in.rst]: is there any chance that is values are empty? if so do it should be specified in here.

File: docs/voxa-event.rst [hunk 10]
```diff
@@ -77,6 +77,74 @@ The ``voxaEvent`` Object
 
     :returns Array: A string array of the platform's supported interfaces
 
+  .. js:function:: getUserInformation()
+
+    Object with user personal information from the platform being used.
+
+    .. code-block:: json
```
Comment (rmberrios) [inline docs/voxa-event.rst]: change the codeblock to javascript, comments are highlighted as an error

File: docs/google-sign-in.rst (hunk 9 above)
Comment (omenocal) [reply docs/google-sign-in.rst]: I'm not totally sure, I have made some tests and I do get values in those variables

File: src/platforms/VoxaPlatform.ts [hunk 11]
```diff
@@ -52,6 +52,10 @@ export interface IVoxaPlatformConfig {
   [key: string]: any;
 }
 
+export interface IDialogFlowPlatformConfig extends IVoxaPlatformConfig {
```
Comment (armonge) [inline src/platforms/VoxaPlatform.ts]: let's move this platform specific declaration to the dialog flow module

File: src/platforms/dialogflow/DialogFlowEvent.ts [hunk 12]
```diff
@@ -26,6 +26,8 @@ import {
 } from "actions-on-google";
 import { Context as AWSLambdaContext } from "aws-lambda";
 import { Context as AzureContext } from "azure-functions-ts-essentials";
+import { OAuth2Client } from "google-auth-library";
```
Comment (armonge) [inline src/platforms/dialogflow/DialogFlowEvent.ts]: we should add this to the package.json requirements

Comment (omenocal) [reply src/platforms/dialogflow/DialogFlowEvent.ts]: @armonge done

**Pull request merged**